x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java

```diff
@@ -358,78 +333,132 @@ private void getEncodedResponse(AsyncExecutionId asyncExecutionId,
     public void getResponse(AsyncExecutionId asyncExecutionId,
                             boolean restoreResponseHeaders,
                             ActionListener<R> listener) {
-        getEncodedResponse(asyncExecutionId, restoreResponseHeaders, ActionListener.wrap(
-            (t) -> listener.onResponse(decodeResponse(t.v1()).withExpirationTime(t.v2())),
-            listener::onFailure
-        ));
+        getResponseFromIndex(asyncExecutionId, restoreResponseHeaders, true, listener);
+    }
+
+    private void getResponseFromIndex(AsyncExecutionId asyncExecutionId,
+                                      boolean restoreResponseHeaders,
+                                      boolean checkAuthentication,
+                                      ActionListener<R> outerListener) {
+        final GetRequest getRequest = new GetRequest(index)
+            .preference(asyncExecutionId.getEncoded())
+            .id(asyncExecutionId.getDocId())
+            .realtime(true);
+        clientWithOrigin.get(getRequest, outerListener.delegateFailure((listener, getResponse) -> {
+            if (getResponse.isExists() == false) {
+                listener.onFailure(new ResourceNotFoundException(asyncExecutionId.getEncoded()));
+                return;
+            }
+            final Counter reservedBytes = Counter.newCounter();
+            try {
+                final R resp = parseResponseFromIndex(
+                    asyncExecutionId, getResponse.getSourceInternal(), restoreResponseHeaders, checkAuthentication, reservedBytes);
+                listener.onResponse(resp);
+            } catch (Exception e) {
+                listener.onFailure(e);
+            } finally {
+                // Release the reserved memory for the response only after the listener has completed its execution.
+                circuitBreaker.addWithoutBreaking(-reservedBytes.get());
+            }
+        }));
+    }
+
+
+    private R parseResponseFromIndex(AsyncExecutionId asyncExecutionId, BytesReference source,
+                                     boolean restoreResponseHeaders, boolean checkAuthentication,
+                                     Counter reservedBytes) {
+        // Reserve an extra buffer as XContentParser can use it to hold parsed values.
+        circuitBreaker.addEstimateBytesAndMaybeBreak(source.length(), "parse xContent of async response");
```

Comment:
should we use here: `circuitBreaker.addWithoutBreaking` instead of `addEstimateBytesAndMaybeBreak`, since we've already parsed xContent?